bmoauth.php

```diff
@@ -0,0 +1,242 @@
+      $post_info = array();
+      try {
+        $post_info = phutil_json_decode(PhabricatorStartup::getRawInput());
+      }
+      catch(Exception $ex) {
+        throw new Exception(pht('Phabricator BMO Authentication failed due to invalid JSON from Bugzilla.'));
+      }
+
+      // Throw exception if either key is not provided by Bugzilla
+      if(!isset($post_info['client_api_key']) || !isset($post_info['client_api_login'])) {
+        throw new Exception(pht('Phabricator BMO Authentication failed due to incomplete JSON from Bugzilla.'));
+      }
+
+      // Generate a transaction code which we'll receive back from Bugzilla
+      // To confirm the API and Client Login information which we saved
+      // In PHP7 this becomes: bin2hex(random_bytes(self::TRANSACTION_CODE_LENGTH))
+      $trans_code = bin2hex(openssl_random_pseudo_bytes(self::TRANSACTION_CODE_LENGTH));
+
+      $csrf = $controller->getExtraURIData();
+
+      // Create a temporary auth token to save the original JS info provided
+      // by the POST from Bugzilla.  Implicitly validates CSRF.
+      $unguarded = AphrontWriteGuard::beginScopedUnguardedWrites();
+      id(new PhabricatorAuthTemporaryToken())
+        ->setTokenResource($trans_code)
+        ->setTokenCode($csrf) // ?
+        ->setTokenType(self::TOKEN_TYPE)
+        ->setTokenExpires(time() + phutil_units('1 hour in seconds'))
+        ->setTemporaryTokenProperty('api_key', $post_info['client_api_key'])
+        ->setTemporaryTokenProperty('client_login', $post_info['client_api_login'])
+        ->save();
+      unset($unguarded);
+
+      // Return a JSON object with a `result` key
+      // This result represents the transaction key we'll use to look up
+      // the JSON information in the next phase of requests
+      $json_obj = array('result' => $trans_code);
+
+      // Render the response JSON
+      $response = id(new AphrontJSONResponse())
+        ->setAddJSONShield(false)
+        ->setHTTPResponseCode(200)
+        ->setContent($json_obj);
+      return array($account, $response);
+    }
+
+    // Verify with CSRF as an additional security measure
+    $this->verifyAuthCSRFCode($request, $controller->getExtraURIData());
+
+    // Match result token and client_api_login to find client_api_key
+    $provided_trans_code = $request->getStr('callback_result');
+    $provided_api_login = $request->getStr('client_api_login');
+    $csrf = $controller->getExtraURIData();
+
+    $token = id(new PhabricatorAuthTemporaryTokenQuery())
+      ->setViewer(PhabricatorUser::getOmnipotentUser())
+      ->withTokenResources(array($provided_trans_code))
+      ->withTokenCodes(array($csrf))
+      ->withTokenTypes(array(self::TOKEN_TYPE))
+      ->withExpired(false)
+      ->executeOne();
+
+    if(!$token) {
+      throw new Exception(pht('No temporary token found for this transaction code (%s) or CSRF token (%s).', $provided_trans_code, $csrf));
```

Comment:
Also, I'm wondering if this is giving too much information to an attacker. We should look into what the proper responses are in the error case.